### ISSUE TITLE
Build-related flags for executables built from source

### ISSUE THREAD

**Issue by Adirio** [kubernetes-sigs/kubebuilder#1401]
# Description

Currently, when `kubebuilder` is built from source, `kubebuilder version` yields:
```
Version: version.Version{KubeBuilderVersion:"unknown", KubernetesVendor:"unknown", GitCommit:"$Format:%H$", BuildDate:"1970-01-01T00:00:00Z", GoOs:"unknown", GoArch:"unknown"}
```

These variables are set as part of the automated build process, but they are never stored in the release branch. they can be found in https://github.com/kubernetes-sigs/kubebuilder/blob/9552d129f7d12cc6c7072a1d323ed9f4b461d550/cmd/version/version.go#L27-L35

# Feature request

Set these flags as part of the development process instead so that, if kubebuilder is built from a release branch, even if it is from source, this fields will be set to provide useful information.

There are three types of variables:

 1. `kubeBuilderVersion`: should be set in the branch.
 2. Those that should be set in the build process:
    1. `goos`: can't be stored in the release branch as there is no one branch per target.
    2. `goarch`: same.
    3. `gitCommit`: the git commit sha1 can't be obtained before commiting the release, so it needs to be set afterwards.
    4. `buildDate`: obvious reasons.
 3. `kubernetesVendorVersion`: has no meaning when building from source.

/kind feature
/kind documentation

**Comment by Adirio** [comment]
The last four variables could be set as part of the `make build` target.

**Comment by mengqiy** [comment]
`kubernetesVendorVersion` indicates the version of the testing tools (kube-apiserver and kubectl) shipped in a release.
It's not useful when building from source.

**Comment by Adirio** [comment]
@estroz [suggested some commands](https://github.com/kubernetes-sigs/kubebuilder/pull/1773#discussion_r520910104) to fetch this data and I would like to expand on them:

- `kubeBuilderVersion`: `git describe --tags --dirty --broken`
- `kubernetesVendorVersion`: I'm not sure if this is currently relevant.
- `goos`: `go env GOOS`
- `goarch`: `go env GOARCH`
- `gitCommit`: `git rev-parse HEAD`
- `buildDate`: `date -u +'%Y-%m-%dT%H:%M:%SZ'` <-- This is not Windows compatible :(

P.S.: I removed the `--always` flag from his suggested `git describe` command as we already include the commit, so no reason to default to the commit if a tag couldn't be found. I also added the `--broken` flag which should hopefully never happen but better safe than sorry.
